### PR TITLE
fix(client): add doctype to preview iframe

### LINKED ISSUE #56660: Different rendered results outside of freecodecamp for Flexbox step 9
### Describe the Issue

In step 9 of the responsive web design: flexbox course, running the code in another browser will produce a different result to running the code in the FCC emulator. I outline the issue in [this forum post](https://forum.freecodecamp.org/t/different-rendered-results-outside-of-freecodecamp-flexbox-photo-gallery-step-9/716021).

I have included the html and css code below. This is the solution to problem 9, and the rendered website in FCC can be seen in the screenshot a).  When I save the code to my computer and run it in my browser, the rendered result looks like screenshot b). Notice the extra white space around the images.
 

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-css-flexbox-by-building-a-photo-gallery/step-9

### Your code

styles.css
```
 * {
  box-sizing: border-box;
}

.gallery {
  border: 5px solid red;
  width: 50%;
}

img {
  width: 100%;
  border: 5px solid blue;
  padding: 5px;
}


```
x.html
```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Photo Gallery</title>
    <link rel="stylesheet" href="./styles.css">
  </head>
  <body>
    <header class="header">
      <h1>css flexbox photo gallery</h1>
    </header>
    <div class="gallery">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/1.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/2.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/3.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/4.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/5.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/6.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/7.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/8.jpg">
      <img src="https://cdn.freecodecamp.org/curriculum/css-photo-gallery/9.jpg">
    </div>
  </body>
</html>


```

### Expected behavior

I expected the rendered output to be the same.

### Screenshots

Image

### System

- Device: [Laptop]
- OS: [Linux Mint 21]
- Browser: [Firefox]
- Version: [131.0.2]

### Additional context

Adding `img {display:block}` in the css fixes the issue. I would guess that somewhere in the FCC rendering, the images are set to block display. This isn't immediately made clear to the user.

### PR DESCRIPTION
**Pull request by lasjorg**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #56660

The preview iframe is running in quirks mode because it's missing the doctype. Example of incorrect output can be found in the issue.

I tried a few things that didn't work, so I ended up adding it to the frame.write(). It's a simple solution and seems to work just fine.

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
While this does fix the specific issue, I don't think we want to force the preview out of quirks mode. If at all possible, we want the preview to exactly represent what the learner has written.

I was just looking into this and it isn't too tricky to make it conditional on the learner having added `<!DOCTYPE html>` in their code.

If it's alright with you, I can push my changes.

**Comment by lasjorg** [comment]
> If at all possible, we want the preview to exactly represent what the learner has written.

The doctype isn't missing in the written code. For example in step 9 of the challenge linked to in the issue a doctype is already in the HTML that was written.

I just think all our previews are missing the doctype not matter what.

**Comment by lasjorg** [comment]
I do get what you are saying. _If_ the camper had missed adding the doctype we would be fixing that. I'm just not sure what else to do. The iframe has to have a doctype and it doesn't get one even if it was added to the HTML in the editor.

BTW, I just noticed that it is also quite broken for the certificate projects. Here we somehow are adding the doctype to the body element.

doctype-placement

**Comment by ojeytonwilliams** [comment]
Yep, that's all correct, to my understanding.

There's a couple of gotchas, but, in short, my fix was to prefix the content with `<!DOCTYPE html>` when the camper has added it. When I tried it, I ended up calling `frame.write(includeDoctype ? '<!DOCTYPE html>' + content : content);` (more or less). `includeDoctype`'s value was conditional on the presence or absence of the doctype, unsurprisingly.

I didn't want to just submit another PR, though, since yours is better than what we have _and_ you figured out what the underlying issue was!

**Comment by lasjorg** [comment]
@ojeytonwilliams How and where are you checking for the doctype?

I moved the code to `writeContentToFrame` and pass `writeToFrame` the doctype if I find it in the `sources.contents`.

File: client/src/templates/Challenges/utils/frame.ts [hunk 1]
```diff
@@ -378,8 +378,15 @@ function writeToFrame(content: string, frame?: FrameDocument) {
 }
 
 const writeContentToFrame = (frameContext: Context) => {
+  let doctype = '';
+
+  if (frameContext.sources.contents) {
+    doctype =
+      frameContext.sources.contents.match(/^<!DOCTYPE html>/i)?.[0] || '';
+  }
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:386]:
```suggestion
  const doctype =
    frameContext.sources.contents?.match(/^<!DOCTYPE html>/i)?.[0] || '';
```
I think the one-liner is clear enough, here. It's a style nit, though, so it doesn't need to block the PR.

**Review by ojeytonwilliams — APPROVED** [review]
My approach was a bit more convoluted, because I'd forgotten that the sources retain the original `<!DOCTYPE html>`, if present. I like your approach a lot more!

File: client/src/templates/Challenges/utils/frame.ts (hunk 1 above)
Comment (lasjorg) [reply client/src/templates/Challenges/utils/frame.ts:386]: The one-liner is cool with me.

**Review by huyenltnguyen — APPROVED** [review]

**Pull request merged**